azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -88,36 +66,138 @@ def _handler_callback(future):
+        logger.debug("HANDLER RUNNER (CLIENT EVENT): Starting runner")
+        _handler_callback = self._generate_callback_for_handler("CLIENT_EVENT")
+
+        # ThreadPool used for running handler functions. By invoking handlers in a separate thread
+        # we can be safe knowing that customer code that has performance issues does not block
+        # client code. Note that the ThreadPool is only used for handler FUNCTIONS (coroutines are
+        # invoked on a dedicated event loop + thread)
+        tpe = concurrent.futures.ThreadPoolExecutor(max_workers=4)
+        event_inbox = self._inbox_manager.get_client_event_inbox()
+        while True:
+            loop = loop_management.get_client_internal_loop()
+            event = await asyncio.wrap_future(
+                asyncio.run_coroutine_threadsafe(event_inbox.get(), loop)
             )
+            if isinstance(event, HandlerRunnerKillerSentinel):
+                # Exit the runner when a HandlerRunnerKillerSentinel is found
+                logger.debug(
+                    "HANDLER RUNNER (CLIENT EVENT): HandlerRunnerKillerSentinel found in event queue. Exiting."
+                )
+                tpe.shutdown()
+                break
+            handler = self._get_handler_for_client_event(event.name)
+            if handler is not None:
+                logger.debug(
+                    "HANDLER RUNNER (CLIENT EVENT): {} event received. Invoking {} handler".format(
+                        event, handler
+                    )
+                )
+                if inspect.iscoroutinefunction(handler):
+                    # Run a coroutine on a dedicated event loop for handler invocations
+                    # TODO: Can we call this on the user loop instead?
+                    handler_loop = loop_management.get_client_handler_loop()
+                    fut = asyncio.run_coroutine_threadsafe(
+                        handler(*event.values_for_user), handler_loop
+                    )
+                    fut.add_done_callback(_handler_callback)
+                else:
+                    # Run a function directly in ThreadPool
+                    fut = tpe.submit(handler, *event.values_for_user)
+                    fut.add_done_callback(_handler_callback)
+            else:
+                logger.debug(
+                    "No handler for event {} set. Skipping handler invocation".format(event)
+                )
 
-        # Schedule a coroutine with the correct type of handler runner
-        inbox = self._get_inbox_for_handler(handler_name)
-        if inbox:
-            coro = self._inbox_handler_runner(inbox, handler_name)
-        else:
-            coro = self._event_handler_runner(handler_name)
+    def _start_handler_runner(self, handler_name):
+        """Create, and store a task for running a handler"""
         # Run the handler runner on a dedicated event loop for handler runners so as to be
         # isolated from all other client activities
```

Comment:
The behavior behind this specific text ("so as to be isolated from all other client activities") bothers me.  Having a dedicated event loop here violates some of the implied contract of async programming.  Our code has been like this for a while, so I don't think we should change it right now, but I think it is a problem.